docs/architecture/adr-003-values-propagation.md

```diff
@@ -0,0 +1,404 @@
+**Value Decision** stage should be independent of value size and 
+incur a roughly constant cost.
+
+In Tendermint, **Value Propagation** is performed via the `PROPOSAL` 
+message, which is broadcast by the designated proposer of the round 
+and includes the proposed value `v`.
+
+The **Value Decision** phase involves `PREVOTE` and `PRECOMMIT` 
+messages—collectively referred to as *votes*. Each vote includes 
+either an identifier `id(v)` of the proposed value `v` or the 
+special value `nil`. The function `id(v)` provides a compact 
+representation of `v`, typically implemented as a fixed-size hash.
+
+From this, we can see that **Value Propagation** is more challenging, 
+as it involves disseminating potentially large amounts of data through 
+the network. In contrast, the **Value Decision** phase requires only 
+the transmission of vote messages, which are relatively small and of 
+constant size. As a result, Malachite’s low-level API provides greater 
+flexibility in the value propagation stage to enable more optimized 
+implementations. 
+
+In this document, we focus on the core parts of the consensus 
+implementation responsible for **Value Propagation**.
+
+TODO: Not sure where to mention this but seems relevant. There have been some slight changes to the Tendermint consensus:
+- `getValue()` is asynchronous and the propose timeout is passed as a parameter in order to let the builder know how long it has to build a value.
+- `valid(v)` is checked when a `PROPOSAL` is available but before the algorithm runs. This is equivalent in a way with `PROPOSAL(h, r, v, vr, valid)` where valid = {Valid | Invalid}.
+- `upon ⟨PROPOSAL, h, r, v, vr⟩ from proposer(h, r)` represents the receiving of a `PROPOSAL` message for a round `r` of height `h` for vale `v`. This is a necessary condition for the success of round `v` with the decision of the proposed value `v`.
+## Current design
+
+### Building Blocks
+_(WIP)_
+
+There are a number of entities that are involved in the value propagation process.
+- Application - this is the software that uses the consensus. It is responsible for:
+  - providing the value to be propagated
+  - validating received values
+  - providing data availability for values (both undecided and decided)
+- Consensus Engine - a component that is responsible for managing the interactions between the consensus core and the application, and between the consensus core and the networking layer. For the ones involved in the value propagation process, the consensus engine is responsible for:
+  - propagating values and/or proposals via the networking layer
+  - receiving values and/or proposals from the networking layer
+  - relaying values and/or proposals to the consensus core
+- Consensus Core - this is the component that implements the tendermint consensus protocol.
+- Networking - this is the component that is responsible for transmitting messages between nodes.
+(Currently not shown in the diagrams)
```

Comment:
Should I add back the `Networking` block?

Reply:
It's fine either way imho, we can leave it like that if you prefer.

Reply:
ok, will leave it out for now